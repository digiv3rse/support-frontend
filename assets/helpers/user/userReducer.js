// @flow

// ----- Imports ----- //

import { type CountryGroupId } from 'helpers/internationalisation/countryGroup';
import { usStates, caStates } from 'helpers/internationalisation/country';

import type { Action } from './userActions';


// ----- Types ----- //

export type User = {
  id: ?string,
  email: string,
  displayName: ?string,
  firstName: string,
  lastName: string,
  fullName: string,
  isTestUser: ?boolean,
  isPostDeploymentTestUser: boolean,
  stateField?: string,
  gnmMarketing: boolean,
  isSignedIn: boolean,
};


// ----- Setup ----- //

const initialState: User = {
  id: '',
  email: '',
  displayName: '',
  firstName: '',
  lastName: '',
  fullName: '',
  isTestUser: null,
  isPostDeploymentTestUser: false,
  gnmMarketing: false,
  isSignedIn: false,
};


// ----- Functions ----- //

function defaultStateOrProvince(countryGroup: CountryGroupId): string {
  if (countryGroup === 'UnitedStates') {
    return Object.keys(usStates)[0];
  } else if (countryGroup === 'Canada') {
    return Object.keys(caStates)[0];
  }

  return '';
}


// ----- Reducer ----- //

function createUserReducer(countryGroup: CountryGroupId) {

  const initialStateWithStateOrProvince = {
    ...initialState,
    stateField: defaultStateOrProvince(countryGroup),
  };

  return function userReducer(
    state: User = initialStateWithStateOrProvince,
    action: Action,
  ): User {

    switch (action.type) {
      case 'SET_USER_ID':
        return { ...state, id: action.id };

      case 'SET_DISPLAY_NAME':
        return { ...state, displayName: action.name };

      case 'SET_FIRST_NAME':
        return { ...state, firstName: action.name };

      case 'SET_LAST_NAME':
        return { ...state, lastName: action.name };

      case 'SET_FULL_NAME':
        return { ...state, fullName: action.name };

      case 'SET_TEST_USER':
        return { ...state, isTestUser: action.testUser };

      case 'SET_POST_DEPLOYMENT_TEST_USER':
        return { ...state, isPostDeploymentTestUser: action.postDeploymentTestUser };

      case 'SET_EMAIL':
        return { ...state, email: action.email };

      case 'SET_STATEFIELD':
        return { ...state, stateField: action.stateField };

      case 'SET_GNM_MARKETING':
        return { ...state, gnmMarketing: action.preference };

      case 'SET_IS_SIGNED_IN':
        return { ...state, isSignedIn: action.isSignedIn };

      default:
        return state;

    }
  };
}

export { createUserReducer };
