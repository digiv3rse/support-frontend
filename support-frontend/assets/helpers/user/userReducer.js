// @flow

// ----- Imports ----- //

import { type CountryGroupId } from 'helpers/internationalisation/countryGroup';

import type { Action } from './userActions';


// ----- Types ----- //

export type User = {
  id: ?string,
  email: string,
  displayName: ?string,
  firstName: string,
  lastName: string,
  fullName: string,
  isTestUser: ?boolean,
  isPostDeploymentTestUser: boolean,
  stateField: string,
  gnmMarketing: boolean,
  isSignedIn: boolean,
  isRecurringContributor: boolean,
  emailValidated: boolean,
};


// ----- Setup ----- //

const initialState: User = {
  id: '',
  email: '',
  displayName: '',
  firstName: '',
  lastName: '',
  fullName: '',
  stateField: '',
  isTestUser: null,
  isPostDeploymentTestUser: false,
  gnmMarketing: false,
  isSignedIn: false,
  isRecurringContributor: false,
  emailValidated: false,
};


// ----- Functions ----- //


// ----- Reducer ----- //

function createUserReducer(countryGroup: CountryGroupId) {

  return function userReducer(
    state: User = initialState,
    action: Action,
  ): User {

    switch (action.type) {
      case 'SET_USER_ID':
        return { ...state, id: action.id };

      case 'SET_DISPLAY_NAME':
        return { ...state, displayName: action.name };

      case 'SET_FIRST_NAME':
        return { ...state, firstName: action.name };

      case 'SET_LAST_NAME':
        return { ...state, lastName: action.name };

      case 'SET_FULL_NAME':
        return { ...state, fullName: action.name };

      case 'SET_TEST_USER':
        return { ...state, isTestUser: action.testUser };

      case 'SET_POST_DEPLOYMENT_TEST_USER':
        return { ...state, isPostDeploymentTestUser: action.postDeploymentTestUser };

      case 'SET_EMAIL':
        return { ...state, email: action.email };

      case 'SET_STATEFIELD':
        return { ...state, stateField: action.stateField };

      case 'SET_GNM_MARKETING':
        return { ...state, gnmMarketing: action.preference };

      case 'SET_IS_SIGNED_IN':
        return { ...state, isSignedIn: action.isSignedIn };

      case 'SET_IS_RECURRING_CONTRIBUTOR':
        return { ...state, isRecurringContributor: true };

      case 'SET_EMAIL_VALIDATED':
        return { ...state, emailValidated: action.emailValidated };

      default:
        return state;
    }
  };
}

export { createUserReducer };
