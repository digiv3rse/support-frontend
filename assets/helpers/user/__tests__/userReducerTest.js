// ----- Imports ----- //

import { createUserReducer } from '../userReducer';


// ----- Tests ----- //

describe('user reducer tests', () => {

  it('should return the initial state', () => {
    expect(createUserReducer('GBPCountries')(undefined, {})).toMatchSnapshot();
  });

  it('should have a default state for the US', () => {
    expect(createUserReducer('UnitedStates')(undefined, {}).stateField).toBe('AK');
  });

  it('should have a default province for Canada', () => {
    expect(createUserReducer('Canada')(undefined, {}).stateField).toBe('AB');
  });

  it('should handle SET_DISPLAY_NAME', () => {
    const name = 'John Doe';
    const action = {
      type: 'SET_DISPLAY_NAME',
      name,
    };

    const newState = createUserReducer('GBPCountries')(undefined, action);

    expect(newState.displayName).toEqual(name);
  });

  it('should handle SET_FIRST_NAME', () => {
    const name = 'John';
    const action = {
      type: 'SET_FIRST_NAME',
      name,
    };

    const newState = createUserReducer('GBPCountries')(undefined, action);
    expect(newState.firstName).toEqual(name);
  });

  it('should handle SET_LAST_NAME', () => {
    const name = 'Doe';
    const action = {
      type: 'SET_LAST_NAME',
      name,
    };

    const newState = createUserReducer('GBPCountries')(undefined, action);

    expect(newState.lastName).toEqual(name);
  });

  it('should handle SET_FULL_NAME', () => {

    const name = 'John Doe';
    const action = {
      type: 'SET_FULL_NAME',
      name,
    };

    const newState = createUserReducer('GBPCountries')(undefined, action);

    expect(newState.fullName).toEqual(name);
  });

  it('should handle SET_TEST_USER', () => {

    const testUser = true;
    const action = {
      type: 'SET_TEST_USER',
      testUser,
    };

    const newState = createUserReducer('GBPCountries')(undefined, action);

    expect(newState.isTestUser).toEqual(testUser);
  });

  it('should handle SET_POST_DEPLOYMENT_TEST_USER', () => {

    const postDeploymentTestUser = true;
    const action = {
      type: 'SET_POST_DEPLOYMENT_TEST_USER',
      postDeploymentTestUser,
    };

    const newState = createUserReducer('GBPCountries')(undefined, action);

    expect(newState.isPostDeploymentTestUser).toEqual(postDeploymentTestUser);
  });

  it('should handle SET_EMAIL', () => {

    const email = 'jdoe@example.com';
    const action = {
      type: 'SET_EMAIL',
      email,
    };

    const newState = createUserReducer('GBPCountries')(undefined, action);

    expect(newState.email).toEqual(email);
  });

  it('should handle SET_STATEFIELD', () => {

    const stateField = 'an american state';
    const action = {
      type: 'SET_STATEFIELD',
      stateField,
    };

    const newState = createUserReducer('UnitedStates')(undefined, action);
    expect(newState.stateField).toEqual(stateField);
  });

  it('should handle SET_GNM_MARKETING', () => {
    const preference = true;

    const action = {
      type: 'SET_GNM_MARKETING',
      preference,
    };

    const newState = createUserReducer('GBPCountries')(undefined, action);
    expect(newState.gnmMarketing).toEqual(preference);
  });
});
