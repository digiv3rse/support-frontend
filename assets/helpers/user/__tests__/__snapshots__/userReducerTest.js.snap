// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user reducer tests should return the initial state 1`] = `
Object {
  "displayName": "",
  "email": "",
  "firstName": "",
  "fullName": "",
  "gnmMarketing": false,
  "id": "",
  "isPostDeploymentTestUser": false,
  "isSignedIn": false,
  "isTestUser": null,
  "lastName": "",
  "stateField": "",
}
`;
